src/drivers/spektrum_rc/spektrum_rc.cpp

```diff
@@ -0,0 +1,304 @@
+
+	// Use a buffer size of the double of the minimum, just to be safe.
+	uint8_t rx_buf[2 * DSM_BUFFER_SIZE];
+
+	// Set up poll topic
+	px4_pollfd_struct_t fds[1];
+	fds[0].fd     = uart_fd;
+	fds[0].events = POLLIN;
+
+	_is_running = true;
+
+	// Main loop
+	while (!_task_should_exit) {
+
+		int pret = px4_poll(&fds[0], (sizeof(fds) / sizeof(fds[0])), 50);
+
+		/* Timed out, do a periodic check for _task_should_exit. */
+		if (pret == 0) {
+			continue;
+		}
+
+		if (pret < 0) {
+			PX4_WARN("poll error %d, %d", pret, errno);
+			/* sleep a bit before next try */
+			usleep(100000);
+			continue;
+		}
+
+		if (!(fds[0].revents & POLLIN)) {
+			// There is only one fd in poll, so this should be impossible.
+			PX4_WARN("poll bit not set");
+			continue;
+		}
+
+		int newbytes = ::read(uart_fd, &rx_buf[0], sizeof(rx_buf));
+
+		if (newbytes < 0) {
+			PX4_WARN("read failed");
+			continue;
+		}
+
+		if (newbytes == 0) {
+			PX4_WARN("read nothing");
+			continue;
+		}
+
+
+		const hrt_abstime now = hrt_absolute_time();
+
+		bool dsm_11_bit;
+		unsigned frame_drops;
+		uint16_t raw_rc_values[input_rc_s::RC_INPUT_MAX_CHANNELS];
+		uint16_t raw_rc_count;
+
+		// parse new data
+		bool rc_updated = dsm_parse(now, rx_buf, newbytes, &raw_rc_values[0], &raw_rc_count,
+					    &dsm_11_bit, &frame_drops, input_rc_s::RC_INPUT_MAX_CHANNELS);
+		UNUSED(dsm_11_bit);
+
+		if (rc_updated) {
+
+			input_rc_s input_rc = {};
+
+			// We don't know RSSI.
+			const int rssi = -1;
+
+			fill_input_rc(raw_rc_count, raw_rc_values, now, false, false, frame_drops, rssi,
+				      input_rc);
+
+			if (rc_pub == nullptr) {
+				rc_pub = orb_advertise(ORB_ID(input_rc), &input_rc);
+
+			} else {
+				orb_publish(ORB_ID(input_rc), rc_pub, &input_rc);
+			}
+		}
+	}
+
+	orb_unadvertise(rc_pub);
+	dsm_deinit();
+
+	_is_running = false;
+}
+
+void fill_input_rc(uint16_t raw_rc_count, uint16_t raw_rc_values[input_rc_s::RC_INPUT_MAX_CHANNELS],
+		   hrt_abstime now, bool frame_drop, bool failsafe, unsigned frame_drops, int rssi,
+		   input_rc_s &input_rc)
+{
+	input_rc.input_source = input_rc_s::RC_INPUT_SOURCE_QURT;
+
+	input_rc.channel_count = raw_rc_count;
+
+	if (input_rc.channel_count > input_rc_s::RC_INPUT_MAX_CHANNELS) {
+		input_rc.channel_count = input_rc_s::RC_INPUT_MAX_CHANNELS;
+	}
+
+	unsigned valid_chans = 0;
+
+	for (unsigned i = 0; i < input_rc.channel_count; ++i) {
+		input_rc.values[i] = raw_rc_values[i];
+
+		if (raw_rc_values[i] != UINT16_MAX) {
+			valid_chans++;
+		}
+	}
+
+	input_rc.timestamp = now;
+	input_rc.timestamp_last_signal = input_rc.timestamp;
+	input_rc.rc_ppm_frame_length = 0;
+
+	/* fake rssi if no value was provided */
+	if (rssi == -1) {
+
+		input_rc.rssi = 255;
+
+	} else {
+		input_rc.rssi = rssi;
+	}
+
+	if (valid_chans == 0) {
+		input_rc.rssi = 0;
+	}
+
+	input_rc.rc_failsafe = failsafe;
+	input_rc.rc_lost = (valid_chans == 0);
+	input_rc.rc_lost_frame_count = frame_drops;
+	input_rc.rc_total_frame_count = 0;
+}
```

Comment:
`rc_ppm_frame_lengh`:
https://github.com/PX4/Firmware/blob/master/msg/input_rc.msg#L25
0 for non-PPM systems

`rc_total_frame_count`:
https://github.com/PX4/Firmware/blob/master/msg/input_rc.msg#L24
We should just increment it but I don't really see the value of this field.